From torvalds@osdl.org  Sat Jun 18 13:15:48 2005
Date: Sat, 18 Jun 2005 13:17:49 -0700 (PDT)
From: Linus Torvalds <torvalds@osdl.org>
To: stable@kernel.org, Chris Wright <chrisw@osdl.org>,
        Greg KH <greg@kroah.com>, Andrew Morton <akpm@osdl.org>
cc: Oleg Nesterov <oleg@tv-sign.ru>, Roland McGrath <roland@redhat.com>,
        Ingo Molnar <mingo@elte.hu>, Steven Rostedt <rostedt@goodmis.org>,
        Kernel security list <security@kernel.org>
Subject: Clean up subthread exec (CAN-2005-1913)
    
Make sure we re-parent itimers.  If subthread exec's with timer pending,
signal is delivered to old group-leader and can panic kernel.

Signed-off-by: Linus Torvalds <torvalds@ppc970.osdl.org>
Signed-off-by: Chris Wright <chrisw@osdl.org>
---
diff --git a/fs/exec.c b/fs/exec.c
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -649,6 +649,7 @@ static inline int de_thread(struct task_
 	}
 	sig->group_exit_task = NULL;
 	sig->notify_count = 0;
+	sig->real_timer.data = (unsigned long)current;
 	spin_unlock_irq(lock);
 
 	/*
